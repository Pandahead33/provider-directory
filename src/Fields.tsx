enum Fields {
  FIRST_NAME = "firstName",
  LAST_NAME = "lastName",
  EMAIL_ADDRESS = "emailAddress",
  SPECIALITY = "speciality",
  PRACTICE_NAME = "practiceName",
}

export default Fields;
