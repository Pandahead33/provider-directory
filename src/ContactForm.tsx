import React, { useState, ReactElement, useEffect } from "react";
import { Contact } from "./Contact";
import Fields from "./Fields";
import Action from "./Action";
import { v4 as uuidv4 } from "uuid";

interface FormProps {
  type: Action;
  actOnProvider: (record: Contact) => void;
  closeForm: () => void;
  initialData: Contact | null | undefined;
}

export default function ContactForm(props: FormProps): ReactElement {
  const firstName = useFormInput("");
  const lastName = useFormInput("");
  const emailAddress = useFormInput("");
  const speciality = useFormInput("N/A");
  const practiceName = useFormInput("");
  const [errorMessages, setErrorMessages] = useState(new Map());

  useEffect(() => {
    firstName.set(props.initialData?.firstName || "");
    lastName.set(props.initialData?.lastName || "");
    emailAddress.set(props.initialData?.emailAddress || "");
    speciality.set(props.initialData?.speciality || "N/A");
    practiceName.set(props.initialData?.practiceName || "");
  }, [props.initialData]);

  const handleSubmit = (event: React.FormEvent<HTMLFormElement>): void => {
    event.preventDefault();
    const target = event.target as HTMLFormElement;

    const fieldValues = new Map([
      [Fields.FIRST_NAME, firstName.value],
      [Fields.LAST_NAME, lastName.value],
      [Fields.EMAIL_ADDRESS, emailAddress.value],
    ]);

    const errorMap = validateFormSubmission(fieldValues);

    if (!target.checkValidity() || errorMap.size > 0) {
      setErrorMessages(errorMap);
      return;
    }

    props.actOnProvider({
      id: props.type === Action.edit ? props.initialData?.id || "" : uuidv4(),
      firstName: firstName.value,
      lastName: lastName.value,
      emailAddress: emailAddress.value,
      speciality: speciality.value,
      practiceName: practiceName.value,
    });

    resetForm();
  };

  const validateFormSubmission = (
    fieldValues: Map<Fields, string>
  ): Map<string, string> => {
    const errorMap: Map<string, string> = new Map(errorMessages);
    const capitalizeFieldName = (fieldName: string) =>
      fieldName.charAt(0).toUpperCase() +
      fieldName.slice(1).replace(/([A-Z])/g, " $1");
    const isValidEmail = (emailValue: string) =>
      /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(emailValue);

    if (!isValidEmail(emailAddress.value)) {
      errorMap.set(
        Fields.EMAIL_ADDRESS,
        `${capitalizeFieldName(
          Fields.EMAIL_ADDRESS
        )} must be in example@gmail.com format.`
      );
    }

    // check if fields are empty
    fieldValues.forEach((fieldValue, fieldName) => {
      if (!fieldValue) {
        errorMap.set(
          fieldName,
          `${capitalizeFieldName(fieldName)} cannot be empty`
        );
      }
    });

    return errorMap;
  };

  const resetForm = (): void => {
    firstName.set("");
    lastName.set("");
    emailAddress.set("");
    speciality.set("N/A");
    practiceName.set("");
    setErrorMessages(new Map());
  };

  const displayedErrorMessages: Array<ReactElement> = [];

  for (const [fieldName, errorMessage] of new Map(errorMessages).entries()) {
    displayedErrorMessages.push(<li key={fieldName}>{errorMessage}</li>);
  }

  return (
    <div className="form">
      <h1>{props.type === Action.add ? "Add Provider" : "Edit Provider"}</h1>
      <div className="error-messages">
        <ul>{displayedErrorMessages}</ul>
      </div>
      <form onSubmit={handleSubmit} noValidate>
        <label htmlFor="firstName">
          <div className="required-field">First Name</div>
          <input
            type="text"
            id="firstName"
            name="firstName"
            value={firstName.value || ""}
            onChange={(event) =>
              firstName.onChange(event, errorMessages, setErrorMessages)
            }
            placeholder="Bob"
            required
          />
        </label>
        <label htmlFor="lastName">
          <div className="required-field">Last Name</div>
          <input
            type="text"
            id="lastName"
            name="lastName"
            value={lastName.value || ""}
            onChange={(event) =>
              lastName.onChange(event, errorMessages, setErrorMessages)
            }
            placeholder="Smith"
            required
          />
        </label>
        <label htmlFor="emailAddress">
          <div className="required-field">Email Address</div>
          <input
            type="email"
            id="emailAddress"
            name="emailAddress"
            value={emailAddress.value || ""}
            onChange={(event) =>
              emailAddress.onChange(event, errorMessages, setErrorMessages)
            }
            placeholder="you@email.com"
            size={55}
            required
          />
        </label>
        <label htmlFor="speciality">
          <div>Speciality</div>
          <select
            id="speciality"
            name="speciality"
            value={speciality.value}
            onChange={(event) =>
              speciality.onChange(event, new Map(), setErrorMessages)
            }
          >
            <option value="N/A">N/A</option>
            <option value="General Medicine">General Medicine</option>
            <option value="Orthopedics">Orthopedics</option>
            <option value="Pediatrics">Pediatrics</option>
            <option value="Podiatry">Podiatry</option>
            <option value="Surgery">Surgery</option>
            <option value="Somnologist">Somnologist</option>
            <option value="Other">Other</option>
          </select>
        </label>
        <label htmlFor="practiceName">
          <div>Practice Name</div>
          <input
            type="text"
            id="practiceName"
            name="practiceName"
            value={practiceName.value || ""}
            onChange={(event) => {
              practiceName.onChange(event, new Map(), setErrorMessages);
            }}
            placeholder="Memorial Family Practice"
            size={55}
          />
        </label>
        <div className="button-container">
          <input
            type="submit"
            value={
              props.type === Action.add ? "âž• Add Provider" : "Edit Provider"
            }
          />
          <input
            className="reset-button"
            type="button"
            value="Reset"
            onClick={resetForm}
          />
          <input
            className="cancel-button"
            type="button"
            value="Cancel"
            onClick={props.closeForm}
          />
        </div>
      </form>
    </div>
  );
}

function useFormInput(initialValue: string) {
  const [value, setValue] = useState(initialValue);

  function handleChange(
    event: React.SyntheticEvent,
    errorMessages: Map<string, string>,
    setErrorMessages: (fieldValue: Map<string, string>) => void | undefined
  ) {
    const target = event.target as HTMLInputElement;
    setValue(target.value);

    switch (target.name) {
      case Fields.FIRST_NAME:
      case Fields.LAST_NAME:
        setErrorMessages(
          validateNameChange(target.name, target.value, errorMessages)
        );
        break;
      case Fields.EMAIL_ADDRESS:
        setErrorMessages(validateEmailChange(target.value, errorMessages));
        break;
      default:
        break;
    }
  }

  return {
    value,
    onChange: handleChange,
    set: setValue,
  };
}

function isValidName(fieldType: Fields, name: string): boolean {
  const containsNumber = (string: string) => /\d/.test(string);
  return !containsNumber(name);
}

function validateNameChange(
  name: Fields,
  value: string,
  errorMessages: Map<string, string>
) {
  const errorMap = new Map(errorMessages);
  if (!isValidName(name, value)) {
    errorMap.set(
      name.toString(),
      `${capitalizeFieldName(name)} should not include numbers`
    );
  } else if (errorMap.has(name.toString())) {
    errorMap.delete(name.toString());
  }

  return errorMap;
}

function isValidEmail(emailAddress: string) {
  const isValidFormat = (emailValue: string) =>
    /^((([!#$%&'*+\-/=?^_`{|}~\w])|([!#$%&'*+\-/=?^_`{|}~\w][!#$%&'*+\-/=?^_`{|}~\.\w]{0,}[!#$%&'*+\-/=?^_`{|}~\w]))[@]\w+([-.]\w+)*\.\w+([-.]\w+)*)$/.test(
      emailValue
    );

  return emailAddress.length === 0 || isValidFormat(emailAddress);
}

function validateEmailChange(email: string, errorMap: Map<string, string>) {
  const fieldType = Fields.EMAIL_ADDRESS.toString();

  if (!isValidEmail(email)) {
    errorMap.set(
      Fields.EMAIL_ADDRESS.toString(),
      `${capitalizeFieldName(
        Fields.EMAIL_ADDRESS
      )} must be in example@gmail.com format.`
    );
  } else if (errorMap.has(fieldType)) {
    errorMap.delete(fieldType);
  }

  return errorMap;
}

function capitalizeFieldName(string: string): string {
  return (
    string.charAt(0).toUpperCase() + string.slice(1).replace(/([A-Z])/g, " $1")
  );
}
