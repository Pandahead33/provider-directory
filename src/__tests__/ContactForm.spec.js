import React from "react"
import {render, screen, fireEvent} from '@testing-library/react'
import { create } from "react-test-renderer";
import ContactForm from "../ContactForm";
import Fields from "../Fields";

test('render and matches snapshot', () => {
    expect(create(<ContactForm />).toJSON()).toMatchSnapshot();
});

test('initial form state', () => {
    render(<ContactForm />);

    const firstNameInput = screen.getByLabelText('First Name');
    const lastNameInput = screen.getByLabelText('Last Name')
    const emailAddressInput = screen.getByLabelText('Email Address');
    const specialityInput = screen.getByLabelText('Speciality');
    const practiceNameInput = screen.getByLabelText('Practice Name');

    expect(firstNameInput.value).toBe('');
    expect(lastNameInput.value).toBe('');
    expect(emailAddressInput.value).toBe('');
    expect(specialityInput.value).toBe('N/A');
    expect(practiceNameInput.value).toBe('');
});

test('reset button clears all form fields', () => {
    render(<ContactForm />);

    const firstNameInput = screen.getByLabelText('First Name');
    const lastNameInput = screen.getByLabelText('Last Name')
    const emailAddressInput = screen.getByLabelText('Email Address');
    const specialityInput = screen.getByLabelText('Speciality');
    const practiceNameInput = screen.getByLabelText('Practice Name');

    firstNameInput.value = "Bob";
    lastNameInput.value = "Smith";
    emailAddressInput.value = "real@gmail.com";
    specialityInput.value = "Somnologist";
    practiceNameInput.value = "A Real Practice MD";
    
    expect(firstNameInput.value).toBe('Bob');
    expect(lastNameInput.value).toBe('Smith');
    expect(emailAddressInput.value).toBe('real@gmail.com');
    expect(specialityInput.value).toBe('Somnologist');
    expect(practiceNameInput.value).toBe('A Real Practice MD');

    fireEvent.click(screen.getByText('Reset'));
    
    expect(firstNameInput.value).toBe('');
    expect(lastNameInput.value).toBe('');
    expect(emailAddressInput.value).toBe('');
    expect(specialityInput.value).toBe('N/A');
    expect(practiceNameInput.value).toBe('');
});

test.each`
    meaning                          | value                          | expectedErrorMessages
    ${"email address has no @"}      | ${"igetspammed1234"}           | ${1}
    ${"email address has no domain"} | ${"igetspammed1234@"}          | ${1}
    ${"email address has no url"}    | ${"igetspammed1234@t."}        | ${1}
    ${"valid email address"}         | ${"igetspammed1234@gmail.com"} | ${0}
`('validate form change where $meaning', async ({field, value, expectedErrorMessages}) => {
    render(<ContactForm />);

    const emailAddressInput = screen.getByLabelText('Email Address');

    fireEvent.change(emailAddressInput, { target: { value: value}});
    
    expect(emailAddressInput.value).toBe(value);

    if (expectedErrorMessages === 1) {
        const errorMessage = await screen.findByText("Email Address must be in example@gmail.com format.");
        expect(errorMessage).toBeTruthy();
    } else {
        const errorMessage = screen.queryByText("Email Address must be in example@gmail.com format.");
        expect(errorMessage).toBeNull();
    }
});

test.each`
    meaning                          | value                          | expectedErrorMessages
    ${"first name includes number"}  | ${"B0b"}                       | ${1}
    ${"valid first name"}            | ${"Bob"}                       | ${0}
`('validate first name form change where $meaning', async ({field, value, expectedErrorMessages}) => {
    render(<ContactForm />);

    const firstNameInput = screen.getByLabelText('First Name');

    fireEvent.change(firstNameInput, { target: { value: value}});
    
    expect(firstNameInput.value).toBe(value);

    if (expectedErrorMessages === 1) {
        const errorMessage = await screen.findByText("First Name should not include numbers");
        expect(errorMessage).toBeTruthy();
    } else {
        const errorMessage = screen.queryByText("First Name should not include numbers`");
        expect(errorMessage).toBeNull();
    }
});

test.each`
    meaning                         | value                            | expectedErrorMessages
    ${"last name includes number"}  | ${"Sm1th"}                       | ${1}
    ${"valid last name"}            | ${"Smith"}                       | ${0}
`('validate first name form change where $meaning', async ({field, value, expectedErrorMessages}) => {
    render(<ContactForm />);

    const lastNameInput = screen.getByLabelText('Last Name');

    fireEvent.change(lastNameInput, { target: { value: value}});
    
    expect(lastNameInput.value).toBe(value);

    if (expectedErrorMessages === 1) {
        const errorMessage = await screen.findByText("Last Name should not include numbers");
        expect(errorMessage).toBeTruthy();
    } else {
        const errorMessage = screen.queryByText("Last Name should not include numbers`");
        expect(errorMessage).toBeNull();
    }
});

test.each`
    scenario                        | field              | errorMessage                                            | invalidValue           | fixedValue
    ${"remove invalid email"}       | ${'Email Address'} | ${"Email Address must be in example@gmail.com format."} | ${"igetspammed1234"}   | ${"igetspammed1234@gmail.com"}
    ${"remove first name has #"}    | ${'First Name'}    | ${"First Name should not include numbers"}              | ${"0"}                 | ${"Bob"}
    ${"remove last name has #"}     | ${'Last Name'}     | ${"Last Name should not include numbers"}               | ${"1"}                 | ${"Smith"}
    ${"empty email to fix invalid"} | ${'Email Address'} | ${"Email Address must be in example@gmail.com format."} | ${"igetspammed1234"}   | ${""}
    ${"empty first name to fix #"}  | ${'First Name'}    | ${"First Name should not include numbers"}              | ${"0"}                 | ${""}
    ${"empty last name to fix #"}   | ${'Last Name'}     | ${"Last Name should not include numbers"}               | ${"1"}                 | ${""}
`('should $scenario error message after fixing', async ({field, invalidValue, errorMessage, fixedValue}) => {
    // insert invalid value for field
    render(<ContactForm />);

    const inputNode = screen.getByLabelText(field);
    
    fireEvent.change(inputNode, { target: { value: invalidValue}});

    // error message should appear 
    const beforeFixErrorMessage = await screen.findByText(errorMessage);
    expect(beforeFixErrorMessage).toBeTruthy();

    // fix e-mail
    fireEvent.change(inputNode, { target: { value: fixedValue}});

    // error message should disappear
    const afterFixErrorMessage = screen.queryByText(errorMessage);
    expect(afterFixErrorMessage).toBeNull();
});

test('validate form submit', () => {

});

test('handleSubmit', () => {

});

test('handleFormChange', () => {

});
