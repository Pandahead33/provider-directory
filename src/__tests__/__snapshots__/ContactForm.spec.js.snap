// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render and matches snapshot 1`] = `
<div
  className="form"
>
  <h1>
    Edit Provider
  </h1>
  <div
    className="error-messages"
  >
    <ul />
  </div>
  <form
    noValidate={true}
    onSubmit={[Function]}
  >
    <label
      htmlFor="firstName"
    >
      <div
        className="required-field"
      >
        First Name
      </div>
      <input
        id="firstName"
        name="firstName"
        onChange={[Function]}
        placeholder="Bob"
        required={true}
        type="text"
        value=""
      />
    </label>
    <label
      htmlFor="lastName"
    >
      <div
        className="required-field"
      >
        Last Name
      </div>
      <input
        id="lastName"
        name="lastName"
        onChange={[Function]}
        placeholder="Smith"
        required={true}
        type="text"
        value=""
      />
    </label>
    <label
      htmlFor="emailAddress"
    >
      <div
        className="required-field"
      >
        Email Address
      </div>
      <input
        id="emailAddress"
        name="emailAddress"
        onChange={[Function]}
        placeholder="you@email.com"
        required={true}
        size={55}
        type="email"
        value=""
      />
    </label>
    <label
      htmlFor="speciality"
    >
      <div>
        Speciality
      </div>
      <select
        id="speciality"
        name="speciality"
        onChange={[Function]}
        value="N/A"
      >
        <option
          value="N/A"
        >
          N/A
        </option>
        <option
          value="General Medicine"
        >
          General Medicine
        </option>
        <option
          value="Orthopedics"
        >
          Orthopedics
        </option>
        <option
          value="Pediatrics"
        >
          Pediatrics
        </option>
        <option
          value="Podiatry"
        >
          Podiatry
        </option>
        <option
          value="Surgery"
        >
          Surgery
        </option>
        <option
          value="Somnologist"
        >
          Somnologist
        </option>
        <option
          value="Other"
        >
          Other
        </option>
      </select>
    </label>
    <label
      htmlFor="practiceName"
    >
      <div>
        Practice Name
      </div>
      <input
        id="practiceName"
        name="practiceName"
        onChange={[Function]}
        placeholder="Memorial Family Practice"
        size={55}
        type="text"
        value=""
      />
    </label>
    <div
      className="button-container"
    >
      <input
        type="submit"
        value="Edit Provider"
      />
      <input
        className="reset-button"
        onClick={[Function]}
        type="button"
        value="Reset"
      />
      <input
        className="cancel-button"
        type="button"
        value="Cancel"
      />
    </div>
  </form>
</div>
`;
